File: sim/events.go
package sim

import "economy/company"

// EventKind categorizes an Event for filtering/display — e.g. a
// research notebook wanting only liquidations, or only fraud
// restatements, across a run.
type EventKind int

const (
	EventFundamental EventKind = iota // a company.FundamentalEvent with Kind != NoJump
	EventRestatement                  // a company.RestatementEvent
	EventLiquidation                  // a forced liquidation order was generated and submitted
	EventIPO                          // an enterprise public listing event
	EventBailout                      // emergency liquidity bailout / syndicate credit line
	EventAcquisition                  // strategic buyout offer at premium
	EventRestructuring                // turnaround cost cutting & operational reorganization
	EventReverseSplit                 // 1-for-10 reverse stock split to maintain listing standards
	EventChapter11                    // court-supervised debt restructuring & clean balance sheet
	EventMacro                        // macroeconomic event (FOMC policy, CPI, labor)
)

func (k EventKind) String() string {
	switch k {
	case EventFundamental:
		return "Fundamental"
	case EventRestatement:
		return "Restatement"
	case EventLiquidation:
		return "Liquidation"
	case EventIPO:
		return "IPO"
	case EventBailout:
		return "Bailout"
	case EventAcquisition:
		return "Acquisition"
	case EventRestructuring:
		return "Restructuring"
	case EventReverseSplit:
		return "ReverseSplit"
	case EventChapter11:
		return "Chapter11"
	case EventMacro:
		return "Macro"
	default:
		return "Unknown"
	}
}

// Event is one notable occurrence during a simulation run, collected
// into Simulation.EventLog. Fields not relevant to a given Kind are
// left at their zero value rather than the log needing a different
// struct per kind — simpler for a first pass at observability, at the
// cost of some unused fields per entry; worth revisiting if the log
// needs to get more structured later.
type Event struct {
	Tick   int
	Kind   EventKind
	Symbol string

	// FundamentalKind/Multiplier populated when Kind == EventFundamental.
	FundamentalKind company.JumpKind
	Multiplier      float64

	// RestatementProfile/PriorGapPercent/Severity populated when
	// Kind == EventRestatement.
	RestatementProfile company.ReportingProfile
	PriorGapPercent    float64
	Severity           float64

	// LiquidatedAgentID/LiquidationQty populated when
	// Kind == EventLiquidation.
	LiquidatedAgentID string
	LiquidationQty    float64

	// IPO details populated when Kind == EventIPO
	IPORevenue float64
	IPOPrice   float64
	IPOShares  float64

	// DistressDetails populated when Kind is a distress event
	DistressDetails string

	// MacroHeadline populated when Kind == EventMacro
	MacroHeadline string
}

File: src/sim/events.rs
use serde::{Deserialize, Serialize};
use crate::company::{JumpKind, ReportingProfile};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum EventKind {
    Fundamental,
    Restatement,
    Liquidation,
    IPO,
    Bailout,
    Acquisition,
    Restructuring,
    ReverseSplit,
    Chapter11,
    Macro,
}

impl EventKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            EventKind::Fundamental => "Fundamental",
            EventKind::Restatement => "Restatement",
            EventKind::Liquidation => "Liquidation",
            EventKind::IPO => "IPO",
            EventKind::Bailout => "Bailout",
            EventKind::Acquisition => "Acquisition",
            EventKind::Restructuring => "Restructuring",
            EventKind::ReverseSplit => "ReverseSplit",
            EventKind::Chapter11 => "Chapter11",
            EventKind::Macro => "Macro",
        }
    }
}

impl std::fmt::Display for EventKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Event {
    pub tick: usize,
    pub kind: EventKind,
    pub symbol: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub fundamental_kind: Option<JumpKind>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub multiplier: Option<f64>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub restatement_profile: Option<ReportingProfile>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prior_gap_percent: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub severity: Option<f64>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub liquidated_agent_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub liquidation_qty: Option<f64>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub ipo_revenue: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ipo_price: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ipo_shares: Option<f64>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub distress_details: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub macro_headline: Option<String>,
}
